soc: st: stm32: add '-align' flag for signing tool v2.21.0+

cdwilson opened the pull request:
Starting in `v2.21.0`, the STM32 signing tool (`STM32_SigningTool_CLI`) stopped automatically adding padding bytes at the beginning of the payload to align it to the `0x400` offset. To restore this behavior, the `-align` flag must be passed to the signing tool post-build command.
    
This PR checks for signing tool version `v2.21.0` or higher and appends the `-align` flag to the post-build signing command. It also changes local CMake variable names to lower case and corrects some indentation issues.
    
Fixes zephyrproject-rtos/zephyr#99456

cdwilson commented:
Set this to draft because I realized this prob needs some updates to the docs. If this looks like the right direction, LMK and I'll update this PR with the doc changes as well.

erwango commented:
So, you can look at [nrf-regtool](https://github.com/zephyrproject-rtos/zephyr/blob/f1c2ce4f4026bd5b212c6e4cc7d754aba69b189e/modules/hal_nordic/nrf-regtool/nrf-regtoolConfigVersion.cmake#L6) which performs a version check once, then store it for later usage.
We should probably do something similar. Are you interested to look at this ?

etienne-lms reviewed (commented):
Worth to add a note in the migration guide to tell flash programming N6 boards require at least STM32_SigningTool_CLI v2.21.0 with a reference to #99456?

cdwilson committed:
soc: st: stm32: add '-align' flag for signing tool v2.21.0+

Starting in v2.21.0, the STM32 signing tool ('STM32_SigningTool_CLI')
stopped automatically adding padding bytes at the beginning of the
payload to align it to the 0x400 offset. To restore this behavior, the
'-align' flag must be passed to the signing tool post-build command.

This commit checks for signing tool version v2.21.0 or higher and
appends the '-align' flag to the post-build signing command. It also
changes local CMake variable names to lower case and corrects some
indentation issues.

Fixes zephyrproject-rtos/zephyr#99456

Signed-off-by: Chris Wilson <chris@binho.io>

cdwilson commented on soc/st/stm32/stm32n6x/CMakeLists.txt:
```diff
@@ -28,20 +28,52 @@ if(NOT CONFIG_BOOTLOADER_MCUBOOT)
+      message(FATAL_ERROR
+      "Unable to determine ${SIGNING_TOOL} version (expected match to "
+      "${SIGNING_TOOL_VERSION_REGEX} regex)"
+      )
```
Not sure if we want to bomb out here or just throw a warning. I figured it's better to have an explicit error than to arbitrarily choose either appending or omitting the `-align` flag in the case where we can't determine the version.

zacck reviewed (commented):
So I just tested this on both 2.20 and 2.21 and it works really smooth

etienne-lms commented on soc/st/stm32/stm32n6x/CMakeLists.txt:
```diff
@@ -22,26 +22,57 @@ zephyr_linker_sources_ifdef(CONFIG_BOOTLOADER_MCUBOOT SECTIONS ram_check.ld)
+          "${signing_tool} version is v${signing_tool_version}, "
+          "appending \"-align\" flag to post-build signing command "
+          "(applicable only with header v2.3 for MCUs, starting from "
+          "version v${signing_tool_version_min})"
```
I think too the message can be omitted. An inline comment should be enough.

erwango commented:
> Adding DNM tag as I'm not able to to run ext flash loaded application on N6-DK using v2.21.0.

Was a local mess up and I was actually missing current fix.

etienne-lms commented on soc/st/stm32/stm32n6x/CMakeLists.txt:
```diff
@@ -22,26 +22,55 @@ zephyr_linker_sources_ifdef(CONFIG_BOOTLOADER_MCUBOOT SECTIONS ram_check.ld)
+        message(STATUS
+          "${signing_tool} version is v${signing_tool_version}, "
+          "appending \"-align\" flag to post-build signing command"
+        )
```
I don't think the message is really useful. Not a blocking issue.

etienne-lms reviewed (commented):
LGTM.
Could you mention in the commit message that you change local CMake variable names to lower case names and correct some indentation issues. For consistency of commit messages to describe changes.

erwango reviewed (approved):
Thanks for this patch!
